Add a fwupd plugin for ASUS ROG ally and ROG ally X

superm1 opened the pull request:
This is only for querying the MCU firmware version for now.  The sequence for writing the firmware update either needs some more RE or some help from ASUS.

@flukejones can I please ask you to test this branch on Ally and Ally X and show the versions it pulls up to see if everything looks sane?

hughsie commented on plugins/asus-hid/asus-hid.quirk:
```diff
@@ -0,0 +1,11 @@
+[USB\VID_0B05&PID_1ABE]
+Plugin = asus_hid
+GType = FuAsusHidDevice
+NumMCU = 2
```
Needs to be namespaced to AsusHidNumMcu -- and be documented in the readme.

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,206 @@
+	guint32 version_command;
+
+	if (mcu == 0)
+		version_command = FU_ASUS_HID_COMMAND_VERSION;
```
I think just stick the fu_struct_asus_hid_command_set_cmd( here

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,206 @@
+		return FALSE;
+
+	part = g_strdup_printf("ASUS\\%s", fu_struct_asus_hid_fw_info_get_product(result));
+	fu_device_add_instance_id(device, part);
```
fu_device_build_instance_id here

hughsie replied:
i.e. `fu_device_add_instance(dev, “PID”,  fu_struct_asus_hid_fw_info_get_product(result)); `

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,206 @@
+	if (!FU_DEVICE_CLASS(fu_asus_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	for (int i = 0; i < self->num_mcu; i++) {
```
guint

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,206 @@
+
+		fu_device_set_version_format(dev_tmp, FWUPD_VERSION_FORMAT_PLAIN);
+		fu_device_set_logical_id(dev_tmp, id);
+		fu_device_add_child(device, dev_tmp);
```
maybe add the child after the version and vendor are set? might be safer...

superm1 commented:
I received an email confirming this is working.  Here was the device ID that shows up:

> ├─N-KEY Device:
> │ │ Device ID: b8a2645298053fb62ea03e27feea6c483d3fd27e
> │ │ Current version: 0.2
> │ │ Vendor: ASUSTek Computer, Inc. (USB:0x0B05)
> │ │ GUID: a42577e5-92b2-53df-a7d1-513812a2ea4f ← USB\VID_0B05&PID_1B4C
> │ │
> │ └─Microcontroller 0:
> │ Device ID: 669b88991d577a8d22a508e5dcbc798726edbf90
> │ Current version: 312
> │ Vendor: ASUSTek Computer, Inc. (USB:0x0B05)
> │ GUID: 9a88fe8b-e444-595e-848e-c1937c54669e ← USB\VID_0B05&PID_1B4C&PART_RC72LA

flukejones commented:
There is an issue where doing the probe now causes the Ally X MCU to start a vibration command and it won't stop until I remove/add my ally driver and then suspend/resume.

This vibration continues even in suspended state.

@superm1 I think the last version that worked fine was a8692e9e6f93 with the Ally X added to quirk

superm1 commented:
> There is an issue where doing the probe now causes the Ally X MCU to start a vibration command and it won't stop until I remove/add my ally driver and then suspend/resume.
> 
> This vibration continues even in suspended state.
> 
> @superm1 I think the last version that worked fine was a8692e9e6f93 with the Ally X added to quirk

Can you try taking out the command that is setting the manufacturer ID? I suspect that's what's doing it and that it happens to do more than we thought

superm1 commented:
> There is an issue where doing the probe now causes the Ally X MCU to start a vibration command and it won't stop until I remove/add my ally driver and then suspend/resume.
> 
> This vibration continues even in suspended state.
> 
> @superm1 I think the last version that worked fine was a8692e9e6f93 with the Ally X added to quirk

Can you try taking out the command that is setting the manufacturer ID? I suspect that's what's doing it and that it happens to do more than we thought

flukejones commented:
It seems I didn't notice this before as the vibration issue occurs on on AC plugged in.  I'll try the suggestion now.

flukejones commented:
@superm1 whcih command is this exactly? I'm not familiar with how fwupd works.

superm1 commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,207 @@
+		fu_device_set_proxy(dev_tmp, device);
+		if (!fu_asus_hid_device_ensure_version(dev_tmp, i, error))
+			return FALSE;
+		if (!fu_asus_hid_device_ensure_manufacturer(dev_tmp, error))
```
@flukejones comment this function call out and see if it helps.

flukejones commented:
Output:
```
├─N-KEY Device:
│ │   Device ID:          b8a2645298053fb62ea03e27feea6c483d3fd27e
│ │   Current version:    0.2
│ │   Vendor:             ASUSTek Computer, Inc. (USB:0x0B05)
│ │   GUID:               a42577e5-92b2-53df-a7d1-513812a2ea4f ← USB\VID_0B05&PID_1B4C
│ │
│ └─Microcontroller 0:
│       Device ID:        669b88991d577a8d22a508e5dcbc798726edbf90
│       Current version:  312_T01
│       Vendor:           ASUSTek Computer, Inc. (USB:0x0B05)
│       GUID:             9a88fe8b-e444-595e-848e-c1937c54669e ← USB\VID_0B05&PID_1B4C&PART_RC72LA
```

Commenting the thing didn't help. But I've done now a full reset on the repo and rebuilt - it seems that another command is resetting the MCU as I'm seeing the LEDs reset to static-mode colour.

superm1 commented:
> Output:
> ```
> ├─N-KEY Device:
> │ │   Device ID:          b8a2645298053fb62ea03e27feea6c483d3fd27e
> │ │   Current version:    0.2
> │ │   Vendor:             ASUSTek Computer, Inc. (USB:0x0B05)
> │ │   GUID:               a42577e5-92b2-53df-a7d1-513812a2ea4f ← USB\VID_0B05&PID_1B4C
> │ │
> │ └─Microcontroller 0:
> │       Device ID:        669b88991d577a8d22a508e5dcbc798726edbf90
> │       Current version:  312_T01
> │       Vendor:           ASUSTek Computer, Inc. (USB:0x0B05)
> │       GUID:             9a88fe8b-e444-595e-848e-c1937c54669e ← USB\VID_0B05&PID_1B4C&PART_RC72LA
> ```
> 
> Commenting the thing didn't help. But I've done now a full reset on the repo and rebuilt - it seems that another command is resetting the MCU as I'm seeing the LEDs reset to static-mode colour.

I wonder if it's actually the opening and closing of the hid endpoint doing this.

flukejones commented:
@superm1 
> I wonder if it's actually the opening and closing of the hid endpoint doing this.

The behaviour I see seems to be consistent for this. From memory it's when the HID endpoint is opened, I have to send a packet to the device to stop vibration when the `hid-asus-ally` driver first loads.

superm1 commented:
> @superm1 
> > I wonder if it's actually the opening and closing of the hid endpoint doing this.
> 
> The behaviour I see seems to be consistent for this. From memory it's when the HID endpoint is opened, I have to send a packet to the device to stop vibration when the `hid-asus-ally` driver first loads.
> 

Can you point me at the packet? I can try to send the same thing on fwupd open.

superm1 commented:
> > @superm1
> > > I wonder if it's actually the opening and closing of the hid endpoint doing this.
> > 
> > 
> > The behaviour I see seems to be consistent for this. From memory it's when the HID endpoint is opened, I have to send a packet to the device to stop vibration when the `hid-asus-ally` driver first loads.
> 
> Can you point me at the packet? I can try to send the same thing on fwupd open.

Is it this stuff?

```c
static int ally_gamepad_init(struct hid_device *hdev, u8 report_id)
{
	const u8 buf[] = { report_id, 0x41, 0x53, 0x55, 0x53, 0x20, 0x54, 0x65,
			   0x63,      0x68, 0x2e, 0x49, 0x6e, 0x63, 0x2e, 0x00 };
	int ret;

	ret = asus_dev_set_report(hdev, buf, sizeof(buf));
	if (ret < 0)
		hid_err(hdev, "Ally failed to send init command: %d\n", ret);

	return ret;
}
```

flukejones commented:
```c
	report->report_id = 0x0D;
	report->ff.enable = 0x0F; /* Enable all by default */
	report->ff.pulse_sustain_10ms = 0xFF; /* Duration */
	report->ff.pulse_release_10ms = 0x00; /* Start Delay */
	report->ff.loop_count = 0xEB; /* Loop Count */
```

I think a HID data packet of `0x0D, 0x0F, 0x00, 0x00, 0x00` should work.

flukejones commented:
> > > @superm1
> Is it this stuff?
> 
> ```c
> static int ally_gamepad_init(struct hid_device *hdev, u8 report_id)
> {
> 	const u8 buf[] = { report_id, 0x41, 0x53, 0x55, 0x53, 0x20, 0x54, 0x65,
> 			   0x63,      0x68, 0x2e, 0x49, 0x6e, 0x63, 0x2e, 0x00 };
> 	int ret;
> 
> 	ret = asus_dev_set_report(hdev, buf, sizeof(buf));
> 	if (ret < 0)
> 		hid_err(hdev, "Ally failed to send init command: %d\n", ret);
> 
> 	return ret;
> }
> ```

No that never stopped the vibration. Have to send the one above. It's possible I'm missing something else in the init but I've never got around to seeing what it is.

superm1 commented:
> I think a HID data packet of 0x0D, 0x0F, 0x00, 0x00, 0x00 should work.

Hmm, I don't see any write to report 13 in the dump you shared me.  Perhaps this means more of the sequence of HID commands before getting the FW version may be needed?

flukejones commented:
Most likely. I am fairly sure there is one that perhaps stops the dev from doing things? Really need an answer from asus about now

superm1 commented:
I spent some time looking at the dll and pulled out some other commands. Will need to see if any of these show up before the get fw version from the dumps.

superm1 commented:
> I have a brief response from ASUS that they are studying update process and will get back to us soon, possibly in the email chain between us and Richard.
> 
> Below are some logs that will help match things up. I think the first one matches the pcap. And I'm willing to bet the bin file bytes would match the pcap data. I'll try to find those bin files again.
> 
> [ROGKBFWUpdateTool_20231105122303.log](https://github.com/user-attachments/files/17126908/ROGKBFWUpdateTool_20231105122303.log) [ROGKBFWUpdateTool_20231105160506.log](https://github.com/user-attachments/files/17126910/ROGKBFWUpdateTool_20231105160506.log)

Can you add a log for Ally X? It seems for Ally that there is a set of flows that does second IC first, flash reset, then first IC and switch to main rom.

And if you can get a packet capture from Ally X (ideally through a VM if you can) it would be good to put it all together.

flukejones commented:
[ally-x-312_t01_update.log](https://github.com/user-attachments/files/17142626/ally-x-312_t01_update.log)

[Full capture data used]. The log matches the pcap here, and the binary data also.

I did spot what looks like a "boot" command, lists the device in the string also as `UUIT8297-BOOT-V1.1` in `Packet 364` in response a `0xD1` packet. Oh yeah, this data looks like exactly what is in the bin file, so `0xD1` must be a "read me this block", and `0xC1` seems like a "write this block", and that data looks like a different format?

For that vibe issue: might be we need to send the `ASUS Tech.Inc.` data (initialise) after reading version. Or after the endpoint is opened before the version request.

**NOTE**: *the data read back in the pcap is the exact data in linked bin file since I forced an overwrite of same version. `0xD1` is hid page, second byte is the block address? I see `FGA80100.RC72LA.312` is also in the bin file as data.

flukejones commented:
I will get a capture of a downgrade to an earlier firmware so we can be double clear on what is what.

superm1 commented:
> [Full capture data used](https://gitlab.com/asus-linux/reverse-engineering/-/tree/master/rog-ally-x/update-capture?ref_type=heads). The log matches the pcap here, and the binary data also.

Man, unfortunately `./contrib/pcap2emulation.py` doesn't want to convert the packet capture to fwupd format.
@fdanis-oss any ideas what is wrong with it?

Good news though, the new firmware parser I put in which pulls the string from the right offset in the binaries seems to be working both on Ally and Ally-X binaries!

```sh
❯ fwupdtool firmware-parse ~/Downloads/FGA80100.RC72LA.312.bin asus-hid
Loading…                 [************************************** ]
<firmware gtype="FuAsusHidFirmware">
  <data size="0x40000">[GInputStream]</data>
  <fga>FGA80100</fga>
  <product>RC72LA</product>
  <version>312</version>
</firmware>
❯ fwupdtool firmware-parse ~/src/asusmcu/ROGMCUFWUpdateTool/fw/FGA80100.RC71LS.318_T02.bin asus-hid
Loading…                 [************************************** ]
<firmware gtype="FuAsusHidFirmware">
  <data size="0x1e000">[GInputStream]</data>
  <fga>FGA80100</fga>
  <product>RC71LS</product>
  <version>318_T02</version>
</firmware>
❯ fwupdtool firmware-parse ~/src/asusmcu/ROGMCUFWUpdateTool/fw/FGA80100.RC71LM.318_T02.bin asus-hid 
Loading…                 [************************************** ]
<firmware gtype="FuAsusHidFirmware">
  <data size="0x1b000">[GInputStream]</data>
  <fga>FGA80100</fga>
  <product>RC71LM</product>
  <version>318_T02</version>
</firmware>
```

> For that vibe issue: might be we need to send the ASUS Tech.Inc. data (initialise) after reading version. Or after the endpoint is opened before the version request.

I've been wondering about that, in the Ghidra stuff and packet capture it looked at first like random stale stack data, but it really does have a static string.  If it's functional then yeah I should be able to force it in to see if it helps.  I'll do that tomorrow.

flukejones commented:
> > For that vibe issue: might be we need to send the ASUS Tech.Inc. data (initialise) after reading version. Or after the endpoint is opened before the version request.
> 
> I've been wondering about that, in the Ghidra stuff and packet capture it looked at first like random stale stack data, but it really does have a static string. If it's functional then yeah I should be able to force it in to see if it helps. I'll do that tomorrow.

I still have the issue in my driver so I gotta sort it out. If I find proper init sequence I'll let you know.

[downgrade from 312 to 311 pcap and log](https://gitlab.com/asus-linux/reverse-engineering/-/tree/master/rog-ally-x/update-capture/downgrade-312-to-311?ref_type=heads)

flukejones commented:
**detach**
```
❯ lsusb
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 002: ID 045e:028e Microsoft Corp. Xbox360 Controller
Bus 001 Device 004: ID 0489:e0f5 Foxconn / Hon Hai Wireless_Device
Bus 001 Device 009: ID 048d:89db Integrated Technology Express, Inc. ITE Upgrade Mode
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 006 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 007 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 008 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
```
**attach**
appears to reset device fine, comes back as:
```
Bus 001 Device 010: ID 0b05:1abe ASUSTek Computer, Inc. N-KEY Device
```

flukejones commented:
For Ally X:
```
❯ fwupdtool detach
Loading…                 [*************************************  ]
WARNING: This package has not been validated, it may not work properly.
Waiting…                 [*************************************  ]
0.	Cancel
1.	aec1a869eb0df71b7cea6b3ac71d39b830faf164 ((null))
2.	75dbd857f37f0f1aa3d5bf2fea25b876288fb415 (AMD Radeon Graphics)
3.	4bde70ba4e39b28f9eab1628f9dd6e6244c03027 (AMD Ryzen Z1 Extreme)
4.	f685512aa07369c9e77742acef941d779d31e766 (GPIO controller)
5.	287b6a8d53a04115c858b26c6a61c24b5a6d745f (Microcontroller 0)
6.	b8a2645298053fb62ea03e27feea6c483d3fd27e (N-KEY Device)
7.	c54ab0237d7a8db8c717b68e0be78e4374a2a079 (Secure Processor)
8.	d1f1d4e16b9688b986c86807a3a4a55df4d1354d (System Firmware)
9.	db0330716216c629bb2c07256e5d018f499eb6ce (System Management Unit (SMU))
10.	1d8d50a4dbc65618f5c399c2ae827b632b3ccc11 (TPM)
11.	6924110cde4fa051bfdc600a60620dc7aa9d3c6a (UEFI Platform Key)
12.	362301da643102b9f38477387e2193e57abaa590 (UEFI dbx)
13.	3743975ad7f64f8d6575a9ae49fb3a8856fe186f (WD PC SN560 SDDPNQE-1T00-1102)
Choose device [0-13]: 6

deck in 🌐 steamfork in fwupd on  superm1/asus-hid is 📦 v2.0.0 via 🐍 v3.11.7 (fwupd) took 6s
❯ lsusb
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 001 Device 003: ID 0489:e0f5 Foxconn / Hon Hai Wireless_Device
Bus 001 Device 004: ID 2808:a753 Generic Focaltech FT9769 ESS
Bus 001 Device 021: ID 048d:89dc Integrated Technology Express, Inc. ITE Upgrade Mode
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 003 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 004 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 005 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 006 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 007 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
Bus 008 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub

deck in 🌐 steamfork in fwupd on  superm1/asus-hid is 📦 v2.0.0 via 🐍 v3.11.7 (fwupd)
❯ fwupdtool attach
Loading…                 [*************************************  ]
WARNING: This package has not been validated, it may not work properly.
Waiting…                 [*************************************  ]
no devices
```

superm1 commented:
Here is the output of the SET reports that occur after the detach (when flashing to the ITE device in update mode).
[reports.txt](https://github.com/user-attachments/files/17184153/reports.txt)

Basically there are 8 commands that occur.
* `0x07` I'm fairly certain is to identify the ITE part
* `0xD1` I'm fairly certain is for dumping the image currently on the SPI part
* `0xC4` is for resetting back from bootloader to app mode.

The remaining commands are:
 * `0xc0`
 * `0xc1`
 * `0xc2`
 * `0xc3`
 * `0xd0`

My thoughts
* 0xC1 feels like an addressing scheme to a remote buffer.  I notice that it always has same ranges it works from.
* 0xC2 only happens twice, in the middle of the backup image dump.
* 0xC0 maybe an erase command, but why so much data in the setup arg then?
* I'm not really sure of the others, if I'm right; presumably 0xD0 or 0xC3 is for flushing the remote buffer.

I was really hoping I'd get lucky and the old SuperIO plugin had the same commands, but they all appear different.

@hughsie you've looked at a lot of dumps and seen the patterns, can you glance through and let me know if anything jumps out to you?

hughsie commented:
> 0xC0 maybe an erase command, but why so much data in the setup arg then?

My guess would be that something is missing a `memset(0)` -- it looks like junk data to me. C0,C1,C2,C3 are used in SCSI (C0=eeprom-read, C1=eeprom-write) but I don't think that's relevant here at all.

antheas commented:
```
❯ sudo hhd.contrib hidraw 2 set:5a070920000100 get:5a set:5a08092000010f set:5a070b260001 get:5a set:5a080b26000102 set:5ae03293
Using argument '2'.
Device: /dev/hidraw2  0b05:1b4c Usage Page: 0xff31 Usage: 0x0080 Names: 'ASUSTeK Computer Inc.': 'N-KEY Device'

Writing provided commands to device:
 - SET 5a070920000100
 - GET 5a
5a07092000010000000000000000000000000000000000000000000000000000
- SET 5a08092000010f
- SET 5a070b260001
 - GET 5a
5a070b2600010200000000000000000000000000000000000000000000000000
 - SET 5a080b26000102
 - SET 5ae03293
```

Device enters setup mode after the last command.

```
sudo hhd.contrib hidraw 0 set:c40000
Using argument '0'.
Device: /dev/hidraw0  048d:89dc Usage Page: 0xff89 Usage: 0x00db Names: 'ITE Tech. Inc.': 'ITE Upgrade Mode'

Writing provided commands to device:
 - SET c40000
```

Device exits setup mode and works as expected.

antheas commented:
> Here is the output of the SET reports that occur after the detach (when flashing to the ITE device in update mode). [reports.txt](https://github.com/user-attachments/files/17184153/reports.txt)

Certain set reports set a memory address in the device, so that the following get report can return information. Wireshark does not support annotating get reports so you might have missed it. It would be worth to also annotate the GET request responses.

superm1 commented:
> > Here is the output of the SET reports that occur after the detach (when flashing to the ITE device in update mode). [reports.txt](https://github.com/user-attachments/files/17184153/reports.txt)
> 
> Certain set reports set a memory address in the device, so that the following get report can return information. Wireshark does not support annotating get reports so you might have missed it. It would be worth to also annotate the GET request responses.

I am ignoring the GET request responses in the tshark CLI for now to build the pattern of everything sent and understand commands, but you're right that for a comprehensive look at it we need to see both.  I was looking at both when I looked in Wireshark GUI.

antheas commented:
It actually takes around 5-10s for the device to enter the bootloader process, so that explains the gap.

```
/dev/hidraw0  048d:89dc Usage Page: 0xff89 Usage: 0x00db Names: 'ITE Tech. Inc.': 'ITE Upgrade Mode'

Device HID Descriptor:
0x06, 0x89, 0xff,  // 000:  Usage Page (Vendor Usage Page 0xff89)
0x09, 0xdb,        // 003:  Usage (Vendor Usage 0xdb)
0xa1, 0x01,        // 005:  Collection (Application)
0x09, 0x01,        // 007:   Usage (Vendor Usage 0x01)
0x15, 0x00,        // 009:    .Logical Minimum (0)
0x26, 0xff, 0x00,  // 00b:    .Logical Maximum (255)
0x75, 0x08,        // 00e:    .Report Size (8)
0x95, 0x3f,        // 010:    .Report Count (63)
0xb1, 0x00,        // 012:   Feature (Data,Arr,Abs)
0xc0,              // 014:  End Collection
```

Very simple HID descriptor. It says the device may receive 63 bytes of information as part of a report and does not specify a report ID. I am not that well versed in HID reports but I would tend to say this means the device may accept any report ID.

superm1 commented:
> Very simple HID descriptor. It says the device may receive 63 bytes of information as part of a report and does not specify a report ID. I am not that well versed in HID reports but I would tend to say this means the device may accept any report ID.

There are no report IDs sent for the flashing sequence and so the first byte of the data is not the report.  It's the command.
It all occurs on "report ID" `0`.

antheas commented:
> > Very simple HID descriptor. It says the device may receive 63 bytes of information as part of a report and does not specify a report ID. I am not that well versed in HID reports but I would tend to say this means the device may accept any report ID.
> 
> There are no report IDs sent for the flashing sequence and so the first byte of the data is not the report. It's the command. It all occurs on "report ID" `0`.

In any case, `sudo hhd.contrib hidraw 0 set:00c40000`,  `sudo hhd.contrib hidraw 0 set:c40000`, and `sudo hhd.contrib hidraw 0 c40000` exit setup mode. So the device understands the commands as report ids as well.

superm1 commented:
> In any case, `sudo hhd.contrib hidraw 0 set:00c40000`, `sudo hhd.contrib hidraw 0 set:c40000`, and `sudo hhd.contrib hidraw 0 c40000` exit setup mode. So the device understands the commands as report ids as well.

If you look more closely at the USB traffic (not the HID layer) you can see that the report ID is explicitly set to 0.  This is because there is no [top level collection](https://learn.microsoft.com/en-us/windows-hardware/drivers/hid/top-level-collections) that includes report IDs.  Even in Windows they [work this way too](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/hidsdi/nf-hidsdi-hidd_setfeature#remarks).

But in any case; no disputes on how to get in and out of bootloader mode.  Both you and Luke have shown the sequence I shared works.  The remaining detail is to interpret the rest of the commands used for the flashing protocol.

antheas commented:
> > In any case, `sudo hhd.contrib hidraw 0 set:00c40000`, `sudo hhd.contrib hidraw 0 set:c40000`, and `sudo hhd.contrib hidraw 0 c40000` exit setup mode. So the device understands the commands as report ids as well.
> 
> If you look more closely at the USB traffic (not the HID layer) you can see that the report ID is explicitly set to 0. This is because there is no [top level collection](https://learn.microsoft.com/en-us/windows-hardware/drivers/hid/top-level-collections) that includes report IDs. Even in Windows they [work this way too](https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/hidsdi/nf-hidsdi-hidd_setfeature#remarks).
> 
> But in any case; no disputes on how to get in and out of bootloader mode. Both you and Luke have shown the sequence I shared works. The remaining detail is to interpret the rest of the commands used for the flashing protocol.

Agreed. If you need anything else lmk.  

I did look into it, and perhaps I should be more stringent with the protocol type, I was just noting that I've yet to find a device where it makes a material difference.

superm1 commented:
I've added support for firmware dumping, specifically while in bootloader mode.  So I would expect that this sequence works:

```sh
fwupdtool detach
fwupdtool get-devices --plugins asus-hid
fwupdtool firmware-dump /tmp/mcu.bin $GUID
fwupdtool firmware-parse /tmp/mcu.bin 
```

The GUID for the firmware-dump command is fetched from the previous command that gets the ID in bootloader mode.
Hopefully the image that is spit out looks conceivably correct and parses correctly.

flukejones commented:
```
deck in 🌐 steamfork in fwupd on  superm1/asus-hid is 📦 v2.0.0 via 🐍 v3.11.7 (fwupd)
❯ fwupdtool detach
Loading…                 [*************************************  ]
WARNING: This package has not been validated, it may not work properly.
Waiting…                 [*************************************  ]
0.	Cancel
1.	aec1a869eb0df71b7cea6b3ac71d39b830faf164 ((null))
2.	75dbd857f37f0f1aa3d5bf2fea25b876288fb415 (AMD Radeon Graphics)
3.	4bde70ba4e39b28f9eab1628f9dd6e6244c03027 (AMD Ryzen Z1 Extreme)
4.	f685512aa07369c9e77742acef941d779d31e766 (GPIO controller)
5.	287b6a8d53a04115c858b26c6a61c24b5a6d745f (Microcontroller 0)
6.	b8a2645298053fb62ea03e27feea6c483d3fd27e (N-KEY Device)
7.	c54ab0237d7a8db8c717b68e0be78e4374a2a079 (Secure Processor)
8.	d1f1d4e16b9688b986c86807a3a4a55df4d1354d (System Firmware)
9.	db0330716216c629bb2c07256e5d018f499eb6ce (System Management Unit (SMU))
10.	1d8d50a4dbc65618f5c399c2ae827b632b3ccc11 (TPM)
11.	6924110cde4fa051bfdc600a60620dc7aa9d3c6a (UEFI Platform Key)
12.	362301da643102b9f38477387e2193e57abaa590 (UEFI dbx)
13.	3743975ad7f64f8d6575a9ae49fb3a8856fe186f (WD PC SN560 SDDPNQE-1T00-1102)
Choose device [0-13]: 6
```
```
deck in 🌐 steamfork in fwupd on  superm1/asus-hid is 📦 v2.0.0 via 🐍 v3.11.7 (fwupd)
❯ fwupdtool get-devices --plugins asus-hid
Loading…                 [************************************** ]
WARNING: This package has not been validated, it may not work properly.
ASUSTeK COMPUTER INC. ROG Ally X RC72LA_RC72LA
│
└─ITE Upgrade Mode:
      Device ID:          b8a2645298053fb62ea03e27feea6c483d3fd27e
      Current version:    0.1
      Vendor:             Integrated Technology Express, Inc. (USB:0x048D)
      GUID:               9c2f0038-4c03-53d0-a1be-f29372c4a69b ← USB\VID_048D&PID_89DC
      Device Flags:       • Is in bootloader mode
                          • Cryptographic hash verification is available
```
```
deck in 🌐 steamfork in fwupd on  superm1/asus-hid is 📦 v2.0.0 via 🐍 v3.11.7 (fwupd)
❯ fwupdtool firmware-dump /tmp/mcu.bin
Loading…                 [*                                      ]
WARNING: This package has not been validated, it may not work properly.
Reading…                 [*                                      ]
Selected device: ITE Upgrade Mode
Reading…                 [***************************************] Less than one minute remaining…
```
verify:
```
Choose firmware [0-109]: 17
<firmware gtype="FuAsusHidFirmware">
  <data size="0x40000">[GInputStream]</data>
  <fga>FGA80100</fga>
  <product>RC72LA</product>
  <version>312_T01</version>
</firmware>
```

The dump + the original flashed FW [here](https://gitlab.com/asus-linux/reverse-engineering/-/tree/master/rog-ally-x/compare-tmp?ref_type=heads)

superm1 commented:
```
Choose firmware [0-109]: 17
<firmware gtype="FuAsusHidFirmware">
  <data size="0x40000">[GInputStream]</data>
  <fga>FGA80100</fga>
  <product>RC72LA</product>
  <version>312_T01</version>
</firmware>
```

Thanks, glad to hear it was able to dump and parse.  Getting further :)

> cmp -l ~/Downloads/mcu.bin ~/Downloads/ROGKBFWUpdateTool/fw/FGA80100.RC72LA.312_T01.bin | gawk '{printf "%08X %02X %02X\n", $1^Cstrtonum(0$2), strtonum(0$3)}' shows that there are differences:

Possible theories are configuration storage and if there is a second bank.  Will need to dig in more closely to decide if there is a mistake.

I have one more very important inquiry; while detached can you upgrade/downgrade a firmware binary from the _Windows tool_?  Because if not, there is more risk on mistakes in the "next step" without a spec.

superm1 commented:
> So the question I think is more: can we recover from a bad flash?

Yes that's exactly right.

> simply booting windows reset it since the drivers are installed for me (and so it will flash)

If necessary, we might be able to run fwupd for _Windows_ to force it in bootloader mode.

superm1 commented:
This link will probably expire, but here is the current artifact from this job for Windows.

https://output.circle-artifacts.com/output/job/6840c2d0-e5b7-47cd-aebd-f7e3d5ddc3d8/artifacts/0/dist/setup/fwupd-2.0.0-setup-x86_64.msi

You should be able to run `fwupdtool.exe detach` from here if we're lucky :)

antheas commented:
> I'm not sure how I would go about testing that.
> 
> Some notes:
> 
> * mcu in boot mode disables wifi (oh, touchscreen disabled too)
> * the mode sticks across reboots and cold boots

Entering bootloader mode does not affect the wifi or touchscreen on my end.

With the following command that is:
`sudo hhd.contrib hidraw 2 set:5a070920000100 get:5a set:5a08092000010f set:5a070b260001 get:5a set:5a080b26000102 set:5ae03293`

I would know since i ran the command through WIFI.

flukejones commented:
> > simply booting windows reset it since the drivers are installed for me (and so it will flash)
> 
> If necessary, we might be able to run fwupd for _Windows_ to force it in bootloader mode.

Ah, something has gone bad in my installed system in such a way that almost everything is disabled on reboot. So discounting this install being trashed, I can say that reboot and cold boot reset the MCU to operating mode.

So what happens if the flash is bad? Is it stuck in boot mode? If so, then I'm sure we can recover fine either by correct flash in Linux or with Windows tool - but I can't verify that until I have enough bad luck.

superm1 commented:
> It doesn't make sense to me that they wouldn't have a way to recover, Windows being what it is. I will ask the guys at ASUS in my next email, and also try to get a clarification of where they are internally.

Yeah I don't doubt _ASUS_ has a way to recover in Windows.  But I'm wondering if it's this tool or an ITE branded/provided tool.  I'll feel so much more warm and fuzzy if it's this one.

> My Ally 1 is standing by for testing 🗡️

OK, you'll be the Guinea pig when we get there :)

> You're making miracles, Mario. Incredible work so far, and much quicker than I would have done.

Thanks!  With the help of IDA I've at least confirmed at least the 0xC0 command garbage data is because they don't clear with `memset()`

![image](https://github.com/user-attachments/assets/6a29bf0c-dd21-4d7c-a6ad-d8d3cac34608)

superm1 commented:
OK so here's the status update.

First the good news.

Thanks to the changes I made for the pcap -> fwupd conversion script (

I can actually get an emulated device now.
```
├─Unknown Device:
│ │   Device ID:          62d5d8280031f607f1db058da959a97f6a8e6d90
│ │   Current version:    0.2
│ │   Vendor:             ASUSTek Computer, Inc. (USB:0x0B05)
│ │   GUIDs:              40e5a299-fe27-5e3b-a4a3-6649aeef4875 ← USB\VID_0B05&PID_1ABE
│ │                       5f0738ae-5e3a-5e94-99fb-62d1b0fefcb1
│ │                       571229d5-caed-5cbc-91c1-5c216601a194
│ │                       8eeac4f3-d98d-57aa-a1fc-b44e8d3d8ba3
│ │                       c33b161d-a538-57df-b1c8-90817ada2387
│ │   Device Flags:       • Emulated
│ │ 
│ ├─Microcontroller 0:
│ │     Device ID:        105a998eae01e451237dc37974beab9362f0066f
│ │     Current version:  311
│ │     Vendor:           ASUSTek Computer, Inc. (USB:0x0B05)
│ │     Update State:     Failed
│ │     Update Error:     failed to write-firmware: writing firmware not supported by device
│ │     Last modified:    2024-10-07
│ │     GUIDs:            f7765461-3ff4-52a0-a4d8-d782b1f885ab ← USB\VID_0B05&PID_1ABE&PART_RC71LS
│ │                       5f0738ae-5e3a-5e94-99fb-62d1b0fefcb1
│ │                       571229d5-caed-5cbc-91c1-5c216601a194
│ │                       8eeac4f3-d98d-57aa-a1fc-b44e8d3d8ba3
│ │                       c33b161d-a538-57df-b1c8-90817ada2387
│ │     Device Flags:     • Updatable
│ │                       • Emulated
│ │   
│ └─Microcontroller 1:
│       Device ID:        c3c241d741fa6d1bac3980e2d79183c6211fc39e
│       Current version:  311
│       Vendor:           ASUSTek Computer, Inc. (USB:0x0B05)
│       GUIDs:            545ddbd6-f041-5d86-bfd7-4764193e0a17 ← USB\VID_0B05&PID_1ABE&PART_RC71LM
│                         5f0738ae-5e3a-5e94-99fb-62d1b0fefcb1
│                         571229d5-caed-5cbc-91c1-5c216601a194
│                         8eeac4f3-d98d-57aa-a1fc-b44e8d3d8ba3
│                         c33b161d-a538-57df-b1c8-90817ada2387
│       Device Flags:     • Updatable
│                         • Emulated

```

Now the bad news:

I have some sort of problem with the device model the way it's being done right now.

```sh
❯ fwupdmgr install /home/supermario/src/asusmcu/ROGMCUFWUpdateTool/fw/FGA80100.RC71LM.318_T02.cab
WARNING: This package has not been validated, it may not work properly.
Writing…                 [                                       ]
failed to write-firmware: writing firmware not supported by device
```

The problem is that I have a parent USB HID device that acts as a proxy to two child devices.  But the updates are "applied" to the children.  I'm expecting that the update basically makes the parent detach, run in IAP/bootloader mode and then the child can proxy commands through it to do the update.  The problem however is that the child is a FuDevice with no write_firmware callback - that I implemented in the parent.  So I either need to make another device class for the children, or create them in the same type as the parent.  I'm not really sure the best way right now.

hughsie commented:
>  So I either need to make another device class for the children

Exactly that, just one that sets up the couple of callbacks to the parent you need -- e.g. https://github.com/fwupd/fwupd/blob/main/plugins/usi-dock/fu-usi-dock-child-device.c

hughsie commented on plugins/asus-hid/asus-hid.quirk:
```diff
@@ -0,0 +1,27 @@
+[USB\VID_048D&PID_89DC]
+Plugin = asus_hid
+Flags = is-bootloader,can-verify-image
+FirmwareSizeMax = 0x40000
```
needs newline here

hughsie commented on plugins/asus-hid/fu-asus-hid-child-device.c:
```diff
@@ -0,0 +1,133 @@
+fu_asus_hid_child_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuAsusHidChildDevice *self = FU_ASUS_HID_CHILD_DEVICE(device);
+	fwupd_codec_string_append(str, idt, "ChipIdx", self->idx ? "1" : "0");
```
Not `fwupd_codec_string_append_int`?

superm1 commented:
Thanks for that style comments, I've fixed it up.  Here's my status update:

I can emulate the device, and after a _lot of hand massaging_ the emulation data I'm at the point that when I pass a test CAB I'm successfully detaching to bootloader, parsing the firmware binary to match the device and keeping the children and gotten up to starting the actual write sequence.

```
05:31:47.579 FuPlugin             superclassed write_firmware(asus_hid)
05:31:47.579 FuMain               Emitting PropertyChanged('Status'='decompressing')
05:31:47.579 FuStruct             FuStructAsusHidDesc:
  fga: FGA80100
  product: RC71LM
  version: 318_T02
05:31:47.579 FuDevice             installing onto c3c241d741fa6d1bac3980e2d79183c6211fc39e:
<firmware gtype="FuAsusHidFirmware">
  <data size="0x1b000">[GInputStream]</data>
  <fga>FGA80100</fga>
  <product>RC71LM</product>
  <version>318_T02</version>
  <firmware>
    <id>payload</id>
    <data size="0x4000">[GInputStream]</data>
  </firmware>
</firmware>

05:31:47.579 FuMain               Emitting PropertyChanged('Status'='device-write')
05:31:47.579 FuHidDevice          HID::SetReport [wValue=0x0300, wIndex=2]:f0 20 00 07 02 
05:31:47.580 FuFirmware           fu_firmware_get_image_by_id: assertion 'error == NULL || *error == NULL' failed
05:31:47.581 FuHistory            modifying device Microcontroller 1 [c3c241d741fa6d1bac3980e2d79183c6211fc39e]
05:31:47.582 FuHistory            modifying device Microcontroller 1 [c3c241d741fa6d1bac3980e2d79183c6211fc39e]
05:31:47.585 FuEngine             failed write-firmware 'failed to send packet: failed to SetReport: no event with ID ControlTransfer:Direction=0x01,RequestType=0x01,Recipient=0x01,Request=0x09,Value=0x0300,Idx=0x0002,Data=8CAABwI=,Length=0x5': FuAsusHidChildDevice:
```
So now I need to continue to massage emulation data for the install phase.

I've posted my latest changes, the emulation data and test cab [here](https://github.com/superm1/ally-emulation) as well if anyone wants to reproduce my result thus far.  Run `make` in that git repo.  Then to use it:

in one terminal:
```sh
fwupd -vv
```
in the other
```sh
fwupdmgr emulation-load ~/ally-emulation/emulation-data/ally.zip
fwupdmgr install ~/ally-emulation/cab/FGA80100.RC71LM.318_T02.cab
```

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,586 @@
+{
+	FuHidDevice *hid_dev;
+
+	/* can be called from child or proxy */
+	if (FU_IS_HID_DEVICE(device))
+		hid_dev = FU_HID_DEVICE(device);
+	else
+		hid_dev = FU_HID_DEVICE(fu_device_get_proxy(device));
+
```
could you use fu_device_get_proxy_with_fallback here?

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,586 @@
+static gboolean
+fu_asus_hid_device_init_seq(FuDevice *device, GError **error)
+{
+	g_autoptr(GByteArray) cmd = fu_struct_asus_hid_command_new();
```
I'd much rather you use the `FuStructAsusHidCommand` typedef here -- then the compiler will do type checking.

hughsie commented on plugins/asus-hid/fu-asus-hid.rs:
```diff
@@ -0,0 +1,165 @@
+    datasz: u8 == 0x6,
+}
+
+// Wire shark filter
+// !(usbhid.setup.ReportID == 13) && _ws.col.protocol == "USBHID" && !(_ws.col.info == "SET_IDLE Request") && !(_ws.col.info == "SET_IDLE Response")
```
maybe stick into the README?

superm1 replied:
I'll drop this when I'm ready to exit draft.  It's just notes for me for now.

superm1 commented:
@hughsie can you maybe guide the right helpers to use?  The payload is tagged in the stream, but every 1024 bytes needs to be written in 17 59 byte chunks followed by one 21 byte chunk.  

This will split it into 59 byte chunks, but that 18th one ends up the wrong size.  So it's almost like I need a chunk array of a chunk array.
```c
	chunks = fu_chunk_array_new_from_stream(stream,
						0,
						FU_STRUCT_ASUS_WRITE_FLASH_COMMAND_SIZE_DATA,
						error);
```

hughsie commented:
> but every 1024 bytes needs to be written in 17 59 byte chunks followed by one 21 byte chunk

Sounds like a ChunkArray of ChunkArray to me :) I think other plugins do something similar, e.g. a 32k "page" and 1 byte "blocks". It might be easier to do use `fu_chunk_array_new_from_stream()` and then `fu_chunk_array_new()` but if you'd like a stream getter for `FuChunk` just yell.

superm1 commented:
> > but every 1024 bytes needs to be written in 17 59 byte chunks followed by one 21 byte chunk
> 
> Sounds like a ChunkArray of ChunkArray to me :) I think other plugins do something similar, e.g. a 32k "page" and 16 byte "blocks". It might be easier to do use `fu_chunk_array_new_from_stream()` and then `fu_chunk_array_new()` but if you'd like a stream getter for `FuChunk` just yell.

Ok!

I think a stream of a stream is more logical than building another array, so yes if you can make a helper for that would be great thanks!

superm1 commented:
I've rebased the plugin against origin/main and I've updated the block/page handling to do a chunk array from stream, and then turn each chunk into it's own chunk array.

I've been focusing currently on Ally as it should be more complex with the two MCU images.  However; all of my work thus far has been off the dump for version "314" on Luke's [R-E repo](https://gitlab.com/asus-linux/reverse-engineering/-/blob/master/rog-ally/firmwareupdate314.pcapng).  I've been slowly updating the [JSON emulation data](https://github.com/superm1/ally-emulation) to throw out what is garbage and what is not against the plugin, but I just had an awful realization that I can't go any further at the moment because there isn't any binaries for the "314" version publicly.  The stuff on the ASUS website is for v318 or v319.

So I'd like to ask for one of these two things to let me keep going:
1. The binaries for version 314.  This is my preference because I don't need to throw away a lot of the work I did on emulation data thus far.
2. A packet capture of an upgrade to version 319.  I can easily package and access v319 from the ASUS website.

flukejones commented:
I unfortunately lost my backup of all previous MCU versions. But it is very easy for someone to get a capture even if they already update.
1. unpack the previous update, find the bin and rename to a number higher than current, flash with the .bat file as admin
2. reboot, and rerun the current update with wireshark captivation

NeroReflex commented:
Capture shared. Cloud recovery did not work and I simply reinstall ed windiws the usual way.
Also yeah usbpcap refused to work in a win to go environment

superm1 commented:
Thanks @NeroReflex!  I've pulled it into my [branch of emulation data](https://github.com/superm1/ally-emulation) and will continue to work with it.

superm1 commented:
With my cleaned up emulation data at https://github.com/superm1/ally-emulation I've managed to do a full write sequence for the microcontroller RC71LS, but not for RC71LM.  
I still need to dig into whether it's an emulation data capture problem or plugin problem with offsets for RC71LM.

@hughsie I was about to rebase, but with the changes in https://github.com/fwupd/fwupd/commit/88050eef680945af4cbed1379a035c7574ee2ce4 I wanted to check how you think this should be handled.  I need to send a few commands at the start of the page, write all the blocks, send a command at end of page.  So walking a straightforward loop doesn't work unless I special case.  Thoughts?  Or maybe if you have an idea just (force) push to the branch.

hughsie commented:
>  So walking a straightforward loop doesn't work unless I special case.

I think you're going to need a FuChunkArray of a FuChunkArray still - you need to do (failable) stuff at the start and end of the page. I've got a branch which adds the fu_chunk_get_stream() API you need, but it depends on the steelseries rustgen branch. If you want a sneak peak it's predictably `hughsie/fu_chunk_stream_new`

superm1 commented:
> > So walking a straightforward loop doesn't work unless I special case.
> 
> I think you're going to need a FuChunkArray of a FuChunkArray still - you need to do (failable) stuff at the start and end of the page. 

OK thanks, that clarifies my big confusion, I didn't see how this API could possibly work for that purpose without a `FuChunkArray` of a `FuChunkArray`.

> I've got a branch which adds the fu_chunk_get_stream() API you need, but it depends on the steelseries rustgen branch. If you want a sneak peak it's predictably `hughsie/fu_chunk_stream_new`

Ack, as I still haven't verified RC71LM, I'll hold off on rebasing.  Maybe by the time I'm ready that will have landed :)

hughsie commented:
@superm1 I'd be fine if you wanted to merge this to avoid it bitrotting. We can always disable it by default in the build if you want.

superm1 commented:
> @superm1 I'd be fine if you wanted to merge this to avoid it bitrotting. We can always disable it by default in the build if you want.

That's not a bad idea to merge as is with some guard rail. I don't mind it being enabled by default, enumeration is definitely correct.

I am thinking to add a force check for the write routine though. Once I find the time to find the problem with the other microcontroller can drop that. I'll spin a change.

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,628 @@
+			return FALSE;
+		if (!fu_asus_hid_device_ensure_version(dev_tmp, i, error))
+			return FALSE;
+		fu_device_add_flag(dev_tmp, FWUPD_DEVICE_FLAG_UPDATABLE);
+		fu_device_add_protocol(dev_tmp, "com.asus.hid");
```
can you add these in the child init?

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,628 @@
+		fu_hid_device_set_interface(FU_HID_DEVICE(device), 2);
+
+	/* FuHidDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_asus_hid_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	return TRUE;
```
just return FU_DEVICE_CLASS(...

hughsie commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,628 @@
+static void
+fu_asus_hid_device_init(FuAsusHidDevice *self)
+{
+	fu_device_add_private_flag(FU_DEVICE(self), FU_DEVICE_PRIVATE_FLAG_NO_AUTO_REMOVE_CHILDREN);
```
do you need the children to outlive the parent?

superm1 replied:
Yup!  Specifically when doing the detach/attach dance the parent disappears.  Without those children you won't have all the details of what payload goes to what child device.

hughsie commented on plugins/asus-hid/fu-asus-hid-plugin.c:
```diff
@@ -0,0 +1,38 @@
+{
+	FuPlugin *plugin = FU_PLUGIN(obj);
+	FuContext *ctx = fu_plugin_get_context(plugin);
+	fu_plugin_add_device_gtype(plugin, FU_TYPE_ASUS_HID_DEVICE);
```
and the child device for coverage?

superm1 replied:
Ah; right.  I was originally going to have an E2E emulation test, but as we're merging sooner it's better to add the explicit coverage for those.

superm1 commented on plugins/asus-hid/fu-asus-hid-device.c:
```diff
@@ -0,0 +1,628 @@
+								 error))
+			return FALSE;
+
+		if (!fu_asus_hid_device_transfer_feature(device,
```
I started to change it and then remembered why I did it this way.  It's a "generic" helper used both for `FuAsusHidDevice` **and** `FuAsusHidChildDevice`.

superm1 replied:
I guess knowing what call path is used with each as the plugin has matured, I'll make a copy of the func for the child to use instead.

superm1 commented:
Thx, let's hold off to merge 1-2 days more.  I need to make sure I didn't break the installation emulation that was working _so far_ with any of these changes for child vs parent ownership.

superm1 commented:
Yup, found some regression.  I've sorted it out and can get the secondary controller updating again.
```
│ ├─Microcontroller 0:
│ │     Device ID:        105a998eae01e451237dc37974beab9362f0066f
│ │     Current version:  319
│ │     Vendor:           ASUSTek Computer, Inc. (USB:0x0B05)
│ │     Update State:     Success
│ │     Last modified:    2024-11-13
│ │     GUIDs:            f7765461-3ff4-52a0-a4d8-d782b1f885ab ← USB\VID_0B05&PID_1ABE&PART_RC71LS
│ │                       5f0738ae-5e3a-5e94-99fb-62d1b0fefcb1
│ │                       571229d5-caed-5cbc-91c1-5c216601a194
│ │                       8eeac4f3-d98d-57aa-a1fc-b44e8d3d8ba3
│ │                       c33b161d-a538-57df-b1c8-90817ada2387
│ │     Device Flags:     • Updatable
│ │                       • Unsigned Payload
│ │                       • Emulated
```

The primary controller update flow is missing from the emulation data but I'm _suspecting_ a conversion issue because it's quite obviously in the wireshark dump.
